packages/foundry/contracts/hooks/SwapDiscountHook/SwapDiscountHook.sol

```diff
@@ -0,0 +1,119 @@
+    address private immutable trustedRouterAddress;
+    uint256 private _shareTokenId = 1;
+
+    constructor(
+        IVault vaultInstance,
+        address _factoryAddress,
+        address _routerAddress,
+        address _campaignFactory,
+        string memory name,
+        string memory symbol
+    ) VaultGuard(vaultInstance) ERC721(name, symbol) Ownable(msg.sender) {
+        allowedFactoryAddress = _factoryAddress;
+        trustedRouterAddress = _routerAddress;
+        discountCampaignFactory = IDiscountCampaignFactory(_campaignFactory);
+    }
+
+    /// @inheritdoc IHooks
+    function onRegister(
+        address factory,
+        address pool,
+        TokenConfig[] memory,
+        LiquidityManagement calldata
+    ) public view override returns (bool) {
+        return factory == allowedFactoryAddress && IBasePoolFactory(factory).isPoolFromFactory(pool);
+    }
+
+    /// @inheritdoc IHooks
+    function getHookFlags() public pure override returns (HookFlags memory hookFlags) {
+        hookFlags.shouldCallAfterSwap = true;
+        return hookFlags;
+    }
+
+    /// @inheritdoc IHooks
+    function onAfterSwap(
+        AfterSwapParams calldata params
+    ) public override onlyVault returns (bool success, uint256 discountedAmount) {
+        (address campaignAddress, ) = discountCampaignFactory.discountCampaigns(params.pool);
+        if (campaignAddress != address(0)) {
+            IDiscountCampaign campaign = IDiscountCampaign(campaignAddress);
+            (bytes32 campaignID, , , , , address rewardToken, address poolAddress, ) = campaign.campaignDetails();
+            if (
+                params.kind == SwapKind.EXACT_IN &&
+                address(params.tokenOut) == rewardToken &&
+                poolAddress == params.pool
+            ) {
+                mint(params, campaign, campaignID);
+            }
+        }
+        return (true, params.amountCalculatedRaw);
+    }
+
+    /**
+     * @notice Apply discount and mint a new token for the user.
+     * @dev This function mints a token for the user after applying the discount and storing relevant swap data.
+     *      The function uses memory for `UserSwapData` to minimize storage operations until the final write to storage.
+     * @param params The parameters of the swap after it has been executed.
+     *        - params.router: The address of the router handling the swap.
+     *        - params.amountCalculatedRaw: The amount swapped by the user.
+     *        - params.pool: The liquidity pool involved in the swap.
+     */
+    function mint(
+        AfterSwapParams calldata params,
+        IDiscountCampaign _campaign,
+        bytes32 _campaignID
+    ) internal nonReentrant {
+        uint256 newTokenId = _shareTokenId++;
+        address user = IRouterCommon(params.router).getSender();
+        _campaign.updateUserDiscountMapping(_campaignID, newTokenId, user, params.amountCalculatedRaw, block.timestamp);
+        _mint(user, newTokenId);
```

Comment:
Interesting to make the swap hook itself an NFT! Minting a new NFT on every swap seems extravagant, though. It would add a fair bit of gas to the critical path, and would get unwieldy for users doing frequent swaps. You could end up with 32 NFTs, and then have to spend the gas to claim them all individually. Probably untenable on mainnet.

The way this is designed, you are rewarding people based on the order of claiming - not the order of swapping - though you do have a cooldown period, so early swappers can claim first. Based on the documentation, this is intended.

But if people are incentivized to claim as soon as possible after swapping, consider simply sending the token "rebate" immediately, vs. going through all this NFT minting and claiming process. That would reward people for swapping early vs. "claiming" early (by essentially doing both simultaneously), but that seems like a valid approach as well. Then we wouldn't need a cooldown period, and the overhead would be just a bit of accounting logic plus a token transfer, and much less onerous for the user.

As it is, an early swapper would have to keep track of the cooldown period, and any who forgot or were busy/distracted could lose the benefit of swapping early if a later swapper claimed first.

You could even enable the swap delta, and return extra tokens right in the swap, though that would prevent this being used on pools that support unbalanced liquidity.

If you want NFTs involved, maybe the Campaign could be an NFT, and you could mint one per user that says they participated in the campaign.

Reply:
@Mubashir-ali-baig